### PR TITLE
Add support for UWP5661 SoC and IVY5661 board

### PR DESCRIPTION
**Pull request by Mani-Sadhasivam**
Hello,

This is the second attempt on adding UNISOC UWP5661 SoC and 96Boards IVY5661 board support to zephyr. The first attempt was made by UNISOC (#10626) but the PR was closed and discontinued. Now, UNISOC has no intention to do the upstream by themselves but they are open for community contributions. Hence with their approval, I'm posting this PR and taken over the codebase from their previous PR. I've added my copyright for files I modified/created.

**About SoC:**

Unisoc UWP5661 is a highly integrated connectivity single chip which offers the lowest BOM in the industry for smart phone, PC, STB, OTT, IoT and automotive applications. This chip includes 2.4GHz and 5GHz WLAN IEEE 802.11 a/b/g/n/ac, Bluetooth 5.0 with supporting high power mode, Direction Finding and Long Range.

**About Board:**

96Boards IVY5661 is one of the IoT boards of the 96Boards family manufactured by UCRobotics. This board features UWP5661 SoC in 96Boards IoT form factor. More information about this board can be found in 96Boards website: https://www.96boards.org/product/ivy5661

The corresponding PR to add HAL is here: https://github.com/zephyrproject-rtos/hal_unisoc/pull/1

### CODE REVIEW

File: boards/arm/96b_ivy5661/doc/index.rst [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/hello_world
+   :board: 96b_ivy5661
+   :goals: build
+
+Flashing
+========
+
+96Boards IVY5661 can be flashed by two methods, one using the ROM
```
Comment (dbkinder) [inline boards/arm/96b_ivy5661/doc/index.rst]: delete this paragraph.  (It's repeated below, with a correction)

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arm/96b_ivy5661/doc/index.rst (hunk 1 above)
Comment (Mani-Sadhasivam) [reply boards/arm/96b_ivy5661/doc/index.rst]: Oops. Will remove it.

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks.

**Comment by pabigot** [comment]
I haven't looked closely at the port, but in preparation for doing so I have a couple questions:
* The sole source for this processor appears to be 96Boards via [Seeed](https://www.seeedstudio.com/Ivy5661-Wi-Fi-and-Bluetooth-IoT-Solution-SoC-SPRD-UWP5661-Cortex-M4-p-2867.html), where it's out of stock.  Is there reason to believe it will be possible for people to get their hands on this hardware?
* #10626 does not include a reason for its being closed, but one blocker was absence of [public documentation](https://github.com/zephyrproject-rtos/zephyr/pull/10626#issuecomment-489516549) on the hardware.  I don't see detailed information on the [product page](https://unisoc.github.io/chips/UWP5661/).  Is hardware reference documentation going to become available?

Having recently spent time dealing with in-tree drivers for sensors that are so past end-of-life that the vendor doesn't even provide documentation for them anymore, I'm concerned about spending effort moving something into Zephyr---and taking on the upstream responsibility for basic maintenance---if no member companies have the hardware and are willing to devote resources to supporting it.

I personally would not want to approve a new platform without being able to verify for myself that it passes basic tests, but if the concerns above are mistaken and I can get hardware I'm happy to take a more in-depth look.

**Comment by Mani-Sadhasivam** [comment]
> I haven't looked closely at the port, but in preparation for doing so I have a couple questions:
> 
> * The sole source for this processor appears to be 96Boards via [Seeed](https://www.seeedstudio.com/Ivy5661-Wi-Fi-and-Bluetooth-IoT-Solution-SoC-SPRD-UWP5661-Cortex-M4-p-2867.html), where it's out of stock.

I believe that this board is also available on Toabao. I'm not sure if it will be available on Seed studio, but I can check with board vendor.

> Is there reason to believe it will be possible for people to get their hands on this hardware?

Please see below comment.

> * #10626 does not include a reason for its being closed, but one blocker was absence of [public documentation](https://github.com/zephyrproject-rtos/zephyr/pull/10626#issuecomment-489516549) on the hardware.  I don't see detailed information on the [product page](https://unisoc.github.io/chips/UWP5661/).  Is hardware reference documentation going to become available?
> 

Yes, Unisoc is working on public documentation for this SoC and it will be available soon. The reason for closing #10626 is that Unisoc has no interest in doing the upstreaming by themselves. But they are happy if someone from the community does it.

> Having recently spent time dealing with in-tree drivers for sensors that are so past end-of-life that the vendor doesn't even provide documentation for them anymore, I'm concerned about spending effort moving something into Zephyr---and taking on the upstream responsibility for basic maintenance---if no member companies have the hardware and are willing to devote resources to supporting it.
> 

This statement doesn't make sense to me. Is Zephyr going to host only the SoC/boards of the member companies? If that's the case, it needs to be written somewhere so that people don't spend time on upstreaming the non-member platforms. And where does the community contribution comes here? This is purely a community contribution and even if the board is not going to be available in future, we should support the existing users. This is what Linux kernel is doing. The developer who submits the SoC/board/driver support takes the sole responsibility of testing regularly with releases and making sure that there is no regression present. Or even if someone reports some regression, then the developer is responsible for fixing it. If he fails to do so for a long time, then the platform can be considered stale and can be removed with notice. And for basic upstreaming maintenance, we only care about build check, didn't we?

> I personally would not want to approve a new platform without being able to verify for myself that it passes basic tests, but if the concerns above are mistaken and I can get hardware I'm happy to take a more in-depth look.

Again, this is not going to work well. Is it a requirement now that the platform support has to be verified by you in order to be upstreamed?

**Comment by pabigot** [comment]
I'm just a contributor, so please don't take anything I say as being anything more than my opinion.

> This statement doesn't make sense to me. Is Zephyr going to host only the SoC/boards of the member companies?

I didn't intend to imply that.  I'm unaware of any clearly defined Zephyr policy and process for determining whether a new platform /device is acceptable or when an old platform/device can be removed.

>  And for basic upstreaming maintenance, we only care about build check, didn't we?

In practice as Zephyr evolves keeping something building means maintainers making source code changes to code in other areas when API or tooling changes, which is risky when the changes can't be verified because nobody has the hardware.  I've been spending a lot of time recently doing this for sensors that use GPIO, so I'm possibly a little too sensitive to the costs and risks of untested code.

> > I personally would not want to approve a new platform without being able to verify for myself that it passes basic tests, but if the concerns above are mistaken and I can get hardware I'm happy to take a more in-depth look.
> 
> Again, this is not going to work well. Is it a requirement now that the platform support has to be verified by you in order to be upstreamed?

Somebody has to approve it; it certainly doesn't have to be me.

I was asked to give my review, and I believe that involves more than a cursory "looks like it might work", which is all I could say without hardware and documentation to evaluate it.

I have no problem with buying an interesting and inexpensive board like this in order to support Zephyr's capabilities, so I wanted to know whether it was going to be available so I could give this contribution the careful attention it deserves.

**Comment by Mani-Sadhasivam** [comment]
@galak ping!

File: boards/arm/96b_ivy5661/96b_ivy5661_defconfig [hunk 2]
```diff
@@ -0,0 +1,19 @@
+CONFIG_SOC_SERIES_UWP566X=y
```
Comment (ioannisg) [inline boards/arm/96b_ivy5661/96b_ivy5661_defconfig]: license header missing

**Comment by ioannisg** [comment]
Needs rebase @Mani-Sadhasivam

File: boards/arm/96b_ivy5661/96b_ivy5661_defconfig (hunk 2 above)
Comment (Mani-Sadhasivam) [reply boards/arm/96b_ivy5661/96b_ivy5661_defconfig]: Added

**Comment by Mani-Sadhasivam** [comment]
@ioannisg Rebased on top of master!

**Comment by Mani-Sadhasivam** [comment]
@galak @ioannisg ping!

File: drivers/interrupt_controller/Kconfig.uwp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	  Enable UNISOC uwp5 interrupt controller.
+
+if UWP_ICTL
+
+config UWP_ICTL_0
+	bool "UNISOC UWP irq Interrupt Controller"
+	help
+	  Enable ICTL irq for UNISOC UWP MCU.
+
+config UWP_ICTL_0_NAME
```
Comment (galak) [inline drivers/interrupt_controller/Kconfig.uwp:21]: We should be getting NAMEs from the devicetree and not Kconfig.

File: drivers/interrupt_controller/Kconfig.uwp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	hex "Parent interrupt number to which UWP_ICTL_0 maps"
+	depends on UWP_ICTL_0
+	default 0x40
+
+config UWP_ICTL_1
+	bool "UNISOC UWP fiq Interrupt Controller"
+	help
+	  Enable ICTL fiq for UNISOC UWP MCU.
+
+config UWP_ICTL_1_NAME
```
Comment (galak) [inline drivers/interrupt_controller/Kconfig.uwp:38]: We should be getting NAMEs from the devicetree and not Kconfig.

File: drivers/interrupt_controller/Kconfig.uwp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	hex "Parent interrupt number to which UWP_ICTL_1 maps"
+	depends on UWP_ICTL_1
+	default 0x60
+
+config UWP_ICTL_2
+	bool "UNISOC UWP AON Interrupt Controller"
+	help
+	  Enable AON ICTL for UNISOC UWP MCU.
+
+config UWP_ICTL_2_NAME
```
Comment (galak) [inline drivers/interrupt_controller/Kconfig.uwp:55]: We should be getting NAMEs from the devicetree and not Kconfig.

File: drivers/interrupt_controller/intc_uwp.c [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+		    uwp_ictl_0_init, &uwp_ictl_0_data, &uwp_ictl_0_config,
+		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+			&uwp_ictl_apis);
+#endif /* CONFIG_UWP_ICTL_0 */
+
+#ifdef CONFIG_UWP_ICTL_1
+static void uwp_config_1_irq(struct device *dev);
+static struct device DEVICE_NAME_GET(uwp_ictl_1);
+static struct uwp_ictl_config uwp_ictl_1_config = {
+	.irq_num = DT_UWP_ICTL_1_IRQ,
```
Comment (galak) [inline drivers/interrupt_controller/intc_uwp.c:133]: Rework this to be DT_IRQN()

File: drivers/interrupt_controller/intc_uwp.c [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+	.isr_table_offset = CONFIG_UWP_ICTL_1_OFFSET,
+	.config_func = uwp_config_1_irq,
+};
+
+static struct uwp_ictl_data uwp_ictl_1_data = {
+	.base_addr = DT_UWP_ICTL_0_BASE + sizeof(struct uwp_intc),
```
Comment (galak) [inline drivers/interrupt_controller/intc_uwp.c:139]: DT_REG_ADDR() instead of DT_UWP_ICTL_0_BASE

File: drivers/interrupt_controller/intc_uwp.c [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+};
+
+static void uwp_config_1_irq(struct device *dev)
+{
+	IRQ_CONNECT(DT_UWP_ICTL_1_IRQ,
```
Comment (galak) [inline drivers/interrupt_controller/intc_uwp.c:144]: DT_IRQN()

File: drivers/interrupt_controller/intc_uwp.c [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+				DT_UWP_ICTL_1_IRQ_PRIO,
```
Comment (galak) [inline drivers/interrupt_controller/intc_uwp.c:145]: DT_IRQ(...., priority)

File: drivers/interrupt_controller/intc_uwp.c [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+				uwp_ictl_isr,
+				DEVICE_GET(uwp_ictl_1), 0);
+	irq_enable(DT_UWP_ICTL_1_IRQ);
```
Comment (galak) [inline drivers/interrupt_controller/intc_uwp.c:148]: DT_IRQN()

File: drivers/serial/uart_uwp.c [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+struct uart_uwp_dev_data_t {
+	u32_t baud_rate;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t cb; /**< Callback function pointer */
+	void *user_data;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static struct uart_uwp_dev_data_t uart_uwp_dev_data_0 = {
+	.baud_rate = DT_UART_UWP_SPEED,
```
Comment (galak) [inline drivers/serial/uart_uwp.c:41]: DT_INST_PROP(0, current_speed)

File: drivers/serial/uart_uwp.c [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.cb = NULL,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static const struct uart_device_config uart_uwp_dev_cfg_0 = {
+	.base = (void *)DT_UART_UWP_BASE,
```
Comment (galak) [inline drivers/serial/uart_uwp.c:48]: DT_INST_REG_ADDR(0)

File: drivers/serial/uart_uwp.c [hunk 13]
```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright (c) 2018, UNISOC Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Comment (galak) [inline drivers/serial/uart_uwp.c:6]: Add #define DT_DRV_COMPAT unisoc_uwp_uart

File: drivers/serial/uart_uwp.c [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+static struct uart_uwp_dev_data_t uart_uwp_dev_data_0 = {
+	.baud_rate = DT_UART_UWP_SPEED,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.cb = NULL,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static const struct uart_device_config uart_uwp_dev_cfg_0 = {
+	.base = (void *)DT_UART_UWP_BASE,
+	.sys_clk_freq = DT_UART_UWP_CLOCK,
```
Comment (galak) [inline drivers/serial/uart_uwp.c:49]: DT_INST_PROP(n, clock_frequency)

File: drivers/serial/uart_uwp.c [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+	struct uart_uwp_dev_data_t * const dev_data = DEV_DATA(dev);
+
+	if (dev_data->cb) {
+		dev_data->cb(dev_data->user_data);
+	}
+}
+
+static void uwp_config_0_irq(struct device *dev)
+{
+	IRQ_CONNECT(DT_UART_UWP_IRQ,
```
Comment (galak) [inline drivers/serial/uart_uwp.c:237]: DT_INST_IRQN(n)

File: soc/arm/unisoc_uwp/uwp566x/dts_fixup.h [hunk 16]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2018, UNISOC Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* SoC level DTS fixup file */
+
+#define DT_NUM_IRQ_PRIO_BITS DT_ARM_V7M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
```
Comment (galak) [inline soc/arm/unisoc_uwp/uwp566x/dts_fixup.h:9]: Remove this file, should not be needed anymore.

**Review by galak — CHANGES_REQUESTED** [review]
Looks mostly good.  Need to make DTS updates to use new macros.  Look at other uart drivers to see how we do per-instance generation now.

Also, it seems like maybe the AON uart should have a different compatible (and thus binding).  Not sure i quite understand how its different from the base uart.

**Comment by carlescufi** [comment]
@Mani-Sadhasivam there seems to be an issue with Kconfig. I will check shippable once it’s done to merge the the unisoc hal PR

**Comment by Mani-Sadhasivam** [comment]
> @Mani-Sadhasivam there seems to be an issue with Kconfig. I will check shippable once it’s done to merge the the unisoc hal PR

@carlescufi Let me fix the Kconfig issue as well.

File: soc/arm/unisoc_uwp/uwp566x/Kconfig.soc [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+
+config SOC_UWP5661
+	bool "uwp5661"
+endchoice
+
+if SOC_UWP5661
+
+config UWP5661_DEBUG
+	bool "Prepend debug header, disabling flash verification"
+	depends on XIP
```
Comment (galak) [inline soc/arm/unisoc_uwp/uwp566x/Kconfig.soc:19]: this doesn't seem correct.  If you depend on XIP than 'default n' if !XIP would never come into play.

File: dts/arm/unisoc/uwp5661.dtsi [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+			current-speed = <115200>;
+			clock-frequency = <26000000>;
+			label = "UART_1";
+			status = "disabled";
+		};
+
+		uart2: uart@40838000 {
+			compatible = "unisoc,uwp-uart";
+			reg = <0x40838000 0x30>;
+			interrupts = <0x0D14 3>;
```
Comment (galak) [inline dts/arm/unisoc/uwp5661.dtsi:87]: looks like you are encoding the Zephyr irq here, you shouldn't need to do that.

File: drivers/interrupt_controller/Kconfig.uwp [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+menuconfig UWP_ICTL
+	bool "UNISOC Interrupt Controller Driver"
+	depends on SOC_FAMILY_UWP
+	depends on MULTI_LEVEL_INTERRUPTS
+	help
+	  Enable UNISOC uwp5 interrupt controller.
+
+if UWP_ICTL
+
+config UWP_ICTL_0
```
Comment (galak) [inline drivers/interrupt_controller/Kconfig.uwp:16]: seems like these shouldn't be numbers, but named by irq purpose.

Comment (Mani-Sadhasivam) [reply drivers/interrupt_controller/Kconfig.uwp:16]: Sorry I don't get it. These are individual instances so should be named as `0, 1`. I'll rename `ICTL_3` to `ICTL_AON`

File: drivers/serial/Kconfig.uwp [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+
+config UART_UWP
+	depends on SOC_SERIES_UWP566X
+	bool "UWP UART driver"
+	default n
+	select SERIAL_HAS_DRIVER
+	select SERIAL_SUPPORT_INTERRUPT
+	depends on SOC_FAMILY_UWP
+	help
+	  This option enables the UWP UART driver, for UART_0.
```
Comment (MaureenHelm) [inline drivers/serial/Kconfig.uwp:14]: Is this an instance config or driver config? We've recently removed driver instance Kconfigs, please see 5a149dfca697e2716d0e96770be1358c9ec0e618

File: drivers/serial/Kconfig.uwp [hunk 21]
```diff
@@ -0,0 +1,34 @@
+# UNISOC UWP UART configuration
+
+# Copyright (c) 2018, UNISOC Incorporated
+# SPDX-License-Identifier: Apache-2.0
+
+config UART_UWP
+	depends on SOC_SERIES_UWP566X
+	bool "UWP UART driver"
+	default n
```
Comment (MaureenHelm) [inline drivers/serial/Kconfig.uwp:9]: You can remove all the `default n` lines, they are redundant

File: drivers/interrupt_controller/intc_uwp.c [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+	return uwp_intc_status(intc);
+}
+
+static const struct irq_next_level_api uwp_ictl_apis = {
+	.intr_enable = uwp_ictl_irq_enable,
+	.intr_disable = uwp_ictl_irq_disable,
+	.intr_get_state = uwp_ictl_irq_get_state,
+};
+
+#ifdef CONFIG_UWP_ICTL_0
```
Comment (MaureenHelm) [inline drivers/interrupt_controller/intc_uwp.c:93]: Please see previous comment about removing instance configs

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by carlescufi** [comment]
> @carlescufi @galak Is there any plan to merge this PR? I just kept on rebasing this for close to 8 months and if we decide not to merge this, I'll just close this.
> 
> Thanks
> 
> PS: There is a CI failure that I'll address if I hear anything about merging.

Yes, I don't see why not. I have now merged https://github.com/zephyrproject-rtos/hal_unisoc/pull/3, please update the revision to 4fdb87c6861a88f2ddf6c9609f7bf7278ddc22e1.

@ioannisg @galak any objections to moving forward with this PR?

**Comment by carlescufi** [comment]
@Mani-Sadhasivam could you please send an email to the devel@ mailing list asking for reviewers for this PR?

File: soc/arm/unisoc_uwp/uwp566x/Kconfig.defconfig.series [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+
+if SOC_SERIES_UWP566X
+
+source "soc/arm/unisoc_uwp/uwp566x/Kconfig.defconfig.uwp5661"
+
+config SOC_SERIES
+	string
+	default "uwp566x"
+
+config TEXT_SECTION_OFFSET
```
Comment (ioannisg) [inline soc/arm/unisoc_uwp/uwp566x/Kconfig.defconfig.series:15]: this is renamed to ROM_START_OFFSET, so needs an update

**Review by ioannisg — CHANGES_REQUESTED** [review]

**Pull request closed without merging**